AzureMonitor: Add support for selecting multiple options when using the equals and not equals dimension filters

aangelisc opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

Adds support for selecting multiple dimension values when using the equals and not equals operators in the Azure Monitor datasource.

**Which issue(s) this PR fixes**:

Fixes #48553 

Also fixes a minor bug where the labels would be cleared on selection.

aangelisc committed:
Add support for multiselect

- Add filters param to Dimensions
- Update existing tests
- Add MultiSelect component
- Add helper function to determine valid options
- Update labels hook to account for custom values
- Update go type
- Add function to build valid filters string

aangelisc committed:
Additional go tests

- Ensure query targets are built correctly

aangelisc committed:
Update DimensionFields frontend test

- Corrently rerender components
- Additional test for multiple labels selection
- Better selection of options in react-select components

aangelisc committed:
Fix lint issue

iwysiu commented on pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource.go:
```diff
@@ -129,10 +129,12 @@ func (e *AzureMonitorDatasource) buildQueries(queries []backend.DataQuery, dsInf
 		} else {
 			for i, filter := range azJSONModel.DimensionFilters {
-				if filter.Operator != "eq" && filter.Filter == "*" {
+				if filter.Operator != "eq" && filter.Filter == "*" && len(filter.Filters) == 0 {
```
`Filters` doesn't get cleared when it switches to `starts with`, so it gets a `400 Bad Request` if you:
1) Set a `eq` operation with 2 values
2) switch to the `starts with` operation

aangelisc replied:
Great catch here, have fixed this 😊

iwysiu reviewed (commented):
I don't think it's necessarily wrong, but it seems weird to have `eq` and `ne` using the same field and `sw` using a different one. Making separate filters for each might make more sense.

aangelisc committed:
Reset filters when operator or dimension changes

aangelisc committed:
Terminology

aangelisc commented:
> I don't think it's necessarily wrong, but it seems weird to have `eq` and `ne` using the same field and `sw` using a different one. Making separate filters for each might make more sense.

I think it's acceptable in this case as the logic for the select is the same across both, there's no real difference between the fields for either operator. I'm also weary of duplicating code if it can be succinct 🤔

aangelisc committed:
Update test

andresmgot commented on pkg/tsdb/azuremonitor/types/types.go:
```diff
@@ -139,9 +140,10 @@ type AzureMonitorJSONQuery struct {
-	Filter    string `json:"filter"`
+	Dimension string   `json:"dimension"`
+	Operator  string   `json:"operator"`
+	Filter    string   `json:"filter"`
+	Filters   []string `json:"filters,omitempty"`
```
it doesn't sound right to me that we have Filter and Filters. Is that necessary? Why is not a Filter a slice of one element? It sounds like you are doing that to support the current queries but if that's the case we would need to:
 - Add a comment saying that using `filter` is deprecated, and migrate here to use `filters` (wherever this struct is being populated).
 - In the QueryEditor, migrate it too so after editing a query, it is now using the list.

Does that make sense? That way we can remove the special handling when Filter is set.

aangelisc replied:
A migration would make the most sense here to support current queries and also implement the new functionality. I'll look into this. This would simplify some things as the `Filter` property is essentially only used now for the `sw` operator.

andresmgot replied:
cool, let me know if you find any issue with that.

aangelisc committed:
Add backend migration

- Update types (deprecate Filter field)
- Add migration logic
- Update tests
- Update dimension filters buliding

aangelisc committed:
Add migration test code

aangelisc committed:
Simplify some logic

aangelisc committed:
Add frontend deprecation notice

aangelisc committed:
Add frontend migration logic and migration tests

aangelisc committed:
Update setting of filter values

aangelisc committed:
Update DimensionFields test

aangelisc committed:
Fix linting issues

andresmgot commented on pkg/tsdb/azuremonitor/types/types.go:
```diff
@@ -139,17 +140,27 @@ type AzureMonitorJSONQuery struct {
-	if a.Filter != "" {
-		filter = a.Filter
+func (a AzureMonitorDimensionFilter) ConstructFiltersString() string {
+	if len(a.Filters) == 1 {
+		return fmt.Sprintf("%v %v '%v'", a.Dimension, a.Operator, a.Filters[0])
+	} else {
```
I don't think you need this `if` (the result would be the same with the `else` logic as far as I can see)

aangelisc replied:
Yep will update this 👍

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts:
```diff
@@ -106,10 +106,12 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<AzureM
           dimension: templateSrv.replace(f.dimension, scopedVars),
           operator: f.operator || 'eq',
-          filter: filter || '*', // send * when empty
+          filter: filter || '*', // send * when empty,
```
I don't think you need to setup a default `filter` since it's now deprecated, the behavior should be moved to the `filters`

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -44,7 +44,11 @@ const useDimensionLabels = (data: PanelData | undefined, query: AzureMonitorQuer
     setDimensionLabels((prevLabels) => {
       const newLabels: DimensionLabels = {};
-      for (const label of Object.keys(labelsObj)) {
+      const currentLabels = Object.keys(labelsObj);
+      if (currentLabels.length === 0) {
+        return prevLabels;
+      }
```
this change seems unrelated to the PR?

aangelisc replied:
This was a bug fix for an issue I noticed with the labels. As it was so short I figured I'd just add it here.

andresmgot replied:
that's fine for small patches :+1: What I usually do with these is either add a comment to my own PR review to highlight it or mention it in the PR description so the reviewer knows (and it's clear is not some leftover from the development)

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -146,16 +170,27 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ data, query, dimensio
+                allowCustomValue
+              />
+            ) : (
+              <Select
```
I am not sure why you can only set one filter when using  `starts with`? Is that an API limitation? (Maybe you can explain that in a comment)

aangelisc replied:
This is an API limitation, will add a comment to explain.

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts:
```diff
@@ -22,6 +22,7 @@ export default function migrateQuery(query: AzureMonitorQuery): AzureMonitorQuer
   workingQuery = migrateToDefaultNamespace(workingQuery);
   workingQuery = migrateMetricsDimensionFilters(workingQuery);
   workingQuery = migrateResourceUri(workingQuery);
+  workingQuery = migrateDimensionFilters(workingQuery);
```
there is already a `migrateMetricsDimensionFilters` above. Can we reuse that method or is there any reason for having a different function here?

aangelisc replied:
I personally think the migrations should be kept independent of each other as it makes it easier to track what changes are being made to the query and when. Happy to discuss this and hear opinions on it 😊

andresmgot replied:
I agree, better to have different methods but here it would be helpful to know what's the difference between them. Maybe renaming would help here: the one in line 23 can be  `migrateDimensionToDimensionFilter` and this one `migrateDimensionFilterToArray`. WDYT?

aangelisc replied:
I think a rename is more appropriate in this case 😊

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.test.ts:
```diff
@@ -115,4 +115,103 @@ describe('AzureMonitor: migrateQuery', () => {
+        })
+      );
+    });
+    it('correctly merges values', () => {
+      const dimensionFilters: AzureMetricDimension[] = [
+        { dimension: 'TestDimension', operator: 'eq', filter: 'testFilter2', filters: ['testFilter'] },
+      ];
```
I think that the behavior should be to ignore `filter` if `filters` is set. There should not be a case in which both are used, queries should be either migrated / new or non-migrated.

andresmgot reviewed (commented):
Thanks for the changes @aangelisc ! I have some more suggestions, let me know what you think.

aangelisc committed:
PR comment updates

- Remove unnecessary if/else condition
- Don't set filter default value as queries should be migrated
- Add comment explaining why sw operator only accepts one value
- Remove unnecessary test for merging of old and new filters

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts:
```diff
@@ -105,12 +105,10 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<AzureM
     const dimensionFilters = (item.dimensionFilters ?? [])
       .filter((f) => f.dimension && f.dimension !== 'None')
       .map((f) => {
-        const filter = templateSrv.replace(f.filter ?? '', scopedVars);
```
can you double check if you can remove this filter? my question is because I don't know if queries that reach this function has been migrated already or not. I believe they will but better to be sure.

aangelisc replied:
A query that gets to this point will be migrated first based on hitting [this line](https://github.com/grafana/grafana/blob/7b9929fffe88d81dd0a5dd1397bbd85e88fb9fe0/public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts#L73) and then [this line](https://github.com/grafana/grafana/blob/7b9929fffe88d81dd0a5dd1397bbd85e88fb9fe0/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L143).

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:
```diff
@@ -181,6 +181,7 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ data, query, dimensio
                 allowCustomValue
               />
             ) : (
+              // The API does not currently allow for multiple starts with clauses to be used.
```
nit:
```suggestion
              // The API does not currently allow for multiple "starts with" clauses to be used.
```

aangelisc committed:
Nit on terminology

Co-authored-by: Andres Martinez Gotor <andres.martinez@grafana.com>

andresmgot reviewed (approved):
LGTM! Just make sure that legacy queries, with a template variable in the filter still work after this :trophy:

aangelisc committed:
Rename migrations for clarity
